### PR TITLE
Pipeline additional options

### PR DESCRIPTION
**Pull request by gitlawr**
1. support reusing global Github config when enabling pipeline, for https://github.com/rancher/rancher/issues/11656
2. add an option to run a specific branch, for https://github.com/rancher/rancher/issues/11665
3. add pipeline validator
4. fix missing triggerUser when manually run
5. fix webhook/cron updates on pipeline updates, for https://github.com/rancher/rancher/issues/11906

### CODE REVIEW

File: pkg/api/customization/pipeline/cluster_pipeline.go [hunk 1]
```diff
@@ -269,3 +284,30 @@ func (h *ClusterPipelineHandler) authAddAccount(clusterPipeline *v3.ClusterPipel
 	}
 	return account, nil
 }
+
+func (h *ClusterPipelineHandler) getGithubConfigCR() (*v3.GithubConfig, error) {
+	authConfigObj, err := h.AuthConfigs.ObjectClient().UnstructuredClient().Get("github", metav1.GetOptions{})
```
Comment (zionwu) [inline pkg/api/customization/pipeline/cluster_pipeline.go]: Why not use cache here?

File: pkg/api/customization/pipeline/pipeline.go [hunk 2]
```diff
@@ -85,7 +126,33 @@ func (h *Handler) run(apiContext *types.APIContext) error {
...
+		return err
+	}
+	if string(requestBytes) != "" {
+		if err := json.Unmarshal(requestBytes, &runPipelineInput); err != nil {
+			return err
+		}
+	}
+	if len(pipeline.Spec.Stages) < 1 ||
+		len(pipeline.Spec.Stages[0].Steps) < 1 ||
+		pipeline.Spec.Stages[0].Steps[0].SourceCodeConfig == nil {
```
Comment (zionwu) [inline pkg/api/customization/pipeline/pipeline.go]: The above condition is the same as Validatitor func, what about putting it in a new func and call it in both places:
```
func validDefinition(pipeline)  bool {
if len(pipeline.Spec.Stages) < 1 ||
		len(pipeline.Spec.Stages[0].Steps) < 1 ||
 		pipeline.Spec.Stages[0].Steps[0].SourceCodeConfig == nil {
   return false
}
return true
}
```

File: pkg/api/customization/pipeline/cluster_pipeline.go (hunk 1 above)
Comment (gitlawr) [reply pkg/api/customization/pipeline/cluster_pipeline.go]: We are getting a `GithubConfig` here, which is a subtype of `AuthConfig`(and stored as an AuthConfig crd). So we need to get the unstructured object of the AuthConfig object.

File: pkg/api/customization/pipeline/pipeline.go (hunk 2 above)
Comment (gitlawr) [reply pkg/api/customization/pipeline/pipeline.go]: updated

**Comment by zionwu** [comment]
LGTM

File: pkg/api/customization/pipeline/cluster_pipeline.go [hunk 3]
```diff
@@ -269,3 +284,30 @@ func (h *ClusterPipelineHandler) authAddAccount(clusterPipeline *v3.ClusterPipel
...
+	u, ok := authConfigObj.(runtime.Unstructured)
+	if !ok {
+		return nil, fmt.Errorf("failed to retrieve GithubConfig, cannot read k8s Unstructured data")
+	}
+	storedGithubConfigMap := u.UnstructuredContent()
+
+	storedGithubConfig := &v3.GithubConfig{}
+	mapstructure.Decode(storedGithubConfigMap, storedGithubConfig)
+
+	metadataMap, ok := storedGithubConfigMap["metadata"].(map[string]interface{})
```
Comment (alena1108) [inline pkg/api/customization/pipeline/cluster_pipeline.go]: is it a safe conversion?

File: pkg/controllers/user/pipeline/controller/pipeline/pipeline.go [hunk 4]
```diff
@@ -67,29 +68,29 @@ func (l *Lifecycle) Create(obj *v3.Pipeline) (*v3.Pipeline, error) {
...
+		}
+		obj.Status.NextStart = nextStart
 	}
 
 	//handle webhook
-	if previous.Spec.TriggerWebhook && previous.Status.WebHookID != "" && !obj.Spec.TriggerWebhook {
-		if err := l.deleteHook(previous); err != nil {
+	if obj.Status.WebHookID != "" && !obj.Spec.TriggerWebhook {
+		if err := l.deleteHook(obj); err != nil {
 			logrus.Errorf("fail to delete previous set webhook")
```
Comment (alena1108) [inline pkg/controllers/user/pipeline/controller/pipeline/pipeline.go]: should we log which object failed to delete? otherwise the message is too generic

Comment (gitlawr) [reply pkg/controllers/user/pipeline/controller/pipeline/pipeline.go]: Sure, do you think it's more appropriate to use warning here?

File: pkg/api/customization/pipeline/cluster_pipeline.go (hunk 3 above)
Comment (gitlawr) [reply pkg/api/customization/pipeline/cluster_pipeline.go]: Yes, it is

File: pkg/controllers/user/pipeline/controller/pipeline/pipeline.go (hunk 4 above)
Comment (gitlawr) [reply pkg/controllers/user/pipeline/controller/pipeline/pipeline.go]: Just updated it

**Pull request merged**